### PR TITLE
Fixed volume relabeling

### LINKED ISSUE #99: storage: relabel test failed: unable to change label in the second time
Hi
The label was successful to set in the first time, but when I want to change the label in the second time, it failed.

environment: RHEL-8.2
```
[root@node1 ~]# lsblk -o NAME,FSTYPE,TYPE
NAME                FSTYPE      TYPE
sr0                             rom
vda                             disk
├─vda1              xfs         part
└─vda2              LVM2_member part
  ├─rhel_node1-root xfs         lvm
  └─rhel_node1-swap swap        lvm
vdb                             disk
vdc                             disk
```
### playbook
```
# cat tests/test-fs-relabel.yml 
---
- hosts: all
  become: true
  vars:
    storage_safe_mode: false
    mount_location: '/opt/test1'

  tasks:
    - include_role:
        name: storage
    - include_tasks: get_unused_disk.yml
      vars:
        max_return: 1

    - name: set label
      include_role:
        name: storage
      vars:
        storage_volumes:
          - name: test1
            type: disk
            mount_point: "{{ mount_location }}"
            fs_type: ext4
            disks: "{{ unused_disks }}"
            fs_label: label

    - include_tasks: verify-role-results.yml

    - name: relabel
      include_role:
        name: storage
      vars:
        storage_volumes:
          - name: test1
            type: disk
            mount_point: "{{ mount_location }}"
            fs_type: ext4
            disks: "{{ unused_disks }}"
            fs_label: relabel

    - include_tasks: verify-role-results.yml

    - name: Clean up
      include_role:
        name: storage
      vars:
        storage_volumes:
          - name: test1
            type: disk
            mount_point: "{{ mount_location }}"
            disks: "{{ unused_disks }}"
            state: absent

    - include_tasks: verify-role-results.yml
```
### log
```
TASK [relabel] ****************************************************************************

TASK [storage : Set version specific variables] *******************************************
ok: [192.168.122.101] => (item=/root/ansible-test/storage/vars/RedHat-8.yml)

TASK [storage : define an empty list of pools to be used in testing] **********************
ok: [192.168.122.101]

TASK [storage : define an empty list of volumes to be used in testing] ********************
ok: [192.168.122.101]

TASK [storage : include the appropriate provider tasks] ***********************************
included: /root/ansible-test/storage/tasks/main-blivet.yml for 192.168.122.101

TASK [storage : get a list of rpm packages installed on host machine] *********************
skipping: [192.168.122.101]

TASK [storage : make sure blivet is available] ********************************************
ok: [192.168.122.101]

TASK [storage : initialize internal facts] ************************************************
ok: [192.168.122.101]

TASK [storage : Apply defaults to pools and volumes [1/6]] ********************************

TASK [storage : Apply defaults to pools and volumes [2/6]] ********************************

TASK [storage : Apply defaults to pools and volumes [3/6]] ********************************

TASK [storage : Apply defaults to pools and volumes [4/6]] ********************************

TASK [storage : Apply defaults to pools and volumes [5/6]] ********************************

TASK [storage : Apply defaults to pools and volumes [6/6]] ********************************
ok: [192.168.122.101] => (item={'name': 'test1', 'type': 'disk', 'mount_point': '/opt/test1', 'fs_type': 'ext4', 'disks': ['vdc'], 'fs_label': 'relabel'})

TASK [storage : debug] ********************************************************************
ok: [192.168.122.101] => {
    "_storage_pools": []
}

TASK [storage : debug] ********************************************************************
ok: [192.168.122.101] => {
    "_storage_volumes": [
        {
            "disks": [
                "vdc"
            ],
            "fs_create_options": "",
            "fs_label": "relabel",
            "fs_overwrite_existing": true,
            "fs_type": "ext4",
            "mount_check": 0,
            "mount_device_identifier": "uuid",
            "mount_options": "defaults",
            "mount_passno": 0,
            "mount_point": "/opt/test1",
            "name": "test1",
            "size": 0,
            "state": "present",
            "type": "disk"
        }
    ]
}

TASK [storage : get required packages] ****************************************************
ok: [192.168.122.101]

TASK [storage : make sure required packages are installed] ********************************
ok: [192.168.122.101]

TASK [storage : manage the pools and volumes to match the specified state] ****************
ok: [192.168.122.101]

TASK [storage : debug] ********************************************************************
ok: [192.168.122.101] => {
    "blivet_output": {
        "actions": [],
        "changed": false,
        "failed": false,
        "leaves": [
            "/dev/vda1",
            "/dev/mapper/rhel_node1-root",
            "/dev/mapper/rhel_node1-swap",
            "/dev/vdb",
            "/dev/vdc",
            "/dev/sr0"
        ],
        "mounts": [
            {
                "dump": 0,
                "fstype": "ext4",
                "opts": "defaults",
                "passno": 0,
                "path": "/opt/test1",
                "src": "UUID=e957f05d-1357-4c1a-a028-4fa02f50aaad",
                "state": "mounted"
            }
        ],
        "packages": [
            "lvm2",
            "xfsprogs",
            "e2fsprogs"
        ],
        "pools": [],
        "volumes": [
            {
                "_device": "/dev/vdc",
                "_mount_id": "UUID=e957f05d-1357-4c1a-a028-4fa02f50aaad",
                "disks": [
                    "vdc"
                ],
                "fs_create_options": "",
                "fs_label": "relabel",
                "fs_overwrite_existing": true,
                "fs_type": "ext4",
                "mount_check": 0,
                "mount_device_identifier": "uuid",
                "mount_options": "defaults",
                "mount_passno": 0,
                "mount_point": "/opt/test1",
                "name": "test1",
                "size": 0,
                "state": "present",
                "type": "disk"
            }
        ]
    }
}

TASK [storage : set the list of pools for test verification] ******************************
ok: [192.168.122.101]

TASK [storage : set the list of volumes for test verification] ****************************
ok: [192.168.122.101]

TASK [storage : manage mounts to match the specified state] *******************************
changed: [192.168.122.101] => (item={'src': 'UUID=e957f05d-1357-4c1a-a028-4fa02f50aaad', 'path': '/opt/test1', 'fstype': 'ext4', 'opts': 'defaults', 'dump': 0, 'passno': 0, 'state': 'mounted'})

TASK [storage : tell systemd to refresh its view of /etc/fstab] ***************************
ok: [192.168.122.101]

TASK [storage : Update facts] *************************************************************
ok: [192.168.122.101]

TASK [include_tasks] **********************************************************************
included: /root/ansible-test/storage/tests/verify-role-results.yml for 192.168.122.101

TASK [Print out pool information] *********************************************************
skipping: [192.168.122.101]

TASK [Print out volume information] *******************************************************
ok: [192.168.122.101] => {
    "_storage_volumes_list": [
        {
            "_device": "/dev/vdc",
            "_mount_id": "UUID=e957f05d-1357-4c1a-a028-4fa02f50aaad",
            "disks": [
                "vdc"
            ],
            "fs_create_options": "",
            "fs_label": "relabel",
            "fs_overwrite_existing": true,
            "fs_type": "ext4",
            "mount_check": 0,
            "mount_device_identifier": "uuid",
            "mount_options": "defaults",
            "mount_passno": 0,
            "mount_point": "/opt/test1",
            "name": "test1",
            "size": 0,
            "state": "present",
            "type": "disk"
        }
    ]
}

TASK [Collect info about the volumes.] ****************************************************
ok: [192.168.122.101]

TASK [Read the /etc/fstab file for volume existence] **************************************
ok: [192.168.122.101]

TASK [Verify the volumes listed in storage_pools were correctly managed] ******************

TASK [Clean up variable namespace] ********************************************************
ok: [192.168.122.101]

TASK [Verify the volumes with no pool were correctly managed] *****************************
[WARNING]: The loop variable 'storage_test_volume' is already in use. You should set the
`loop_var` value in the `loop_control` option for the task to something else to avoid
variable collisions and unexpected behavior.
included: /root/ansible-test/storage/tests/test-verify-volume.yml for 192.168.122.101

TASK [set_fact] ***************************************************************************
ok: [192.168.122.101]

TASK [include_tasks] **********************************************************************
included: /root/ansible-test/storage/tests/test-verify-volume-mount.yml for 192.168.122.101
included: /root/ansible-test/storage/tests/test-verify-volume-fstab.yml for 192.168.122.101
included: /root/ansible-test/storage/tests/test-verify-volume-fs.yml for 192.168.122.101
included: /root/ansible-test/storage/tests/test-verify-volume-device.yml for 192.168.122.101

TASK [Set some facts] *********************************************************************
ok: [192.168.122.101]

TASK [Verify the current mount state by device] *******************************************
ok: [192.168.122.101] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Verify the current mount state by mount point] **************************************
ok: [192.168.122.101] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Verify the mount fs type] ***********************************************************
ok: [192.168.122.101] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Unset facts] ************************************************************************
ok: [192.168.122.101]

TASK [Set some variables for fstab checking] **********************************************
ok: [192.168.122.101]

TASK [Verify that the device identifier appears in /etc/fstab] ****************************
ok: [192.168.122.101] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Verify the fstab mount point] *******************************************************
ok: [192.168.122.101] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Clean up variables] *****************************************************************
ok: [192.168.122.101]

TASK [Verify fs type] *********************************************************************
ok: [192.168.122.101] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Verify fs label] ********************************************************************
fatal: [192.168.122.101]: FAILED! => {
    "assertion": false,
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}

PLAY RECAP ********************************************************************************
192.168.122.101            : ok=105  changed=3    unreachable=0    failed=1    skipped=26   rescued=0    ignored=0   

```

### PR DESCRIPTION
**Pull request by japokorn**
- modifying volume label no longer causes test failure
- added forced udev update which should fix occasional test failures
- added test

Fixes #99

### CODE REVIEW

File: library/blivet.py [hunk 1]
```diff
@@ -411,6 +413,11 @@ def _reformat(self):
         """ Schedule actions as needed to ensure the volume is formatted as specified. """
         fmt = self._get_format()
         if self._device.format.type == fmt.type:
+            # format is the same, not need to run reformatting
+            if self._device.format.label != fmt.label and fmt.label != '':
```
Comment (dwlehman) [inline library/blivet.py:417]: We should probably allow clearing/removing a label, no?

**Review by dwlehman — APPROVED** [review]
It looks okay to me. I wonder if that CI failure is the same one I have been hitting. I was seeing udev fail to pick up the label on a new volume. Here is what I'm using now to work around it: https://github.com/linux-system-roles/storage/pull/59/commits/125def8a921fb81f8d86b27f66888916488c7a4b. You could give that a try (first, add `or action.is_configure` to the conditional in the handler).

File: library/blivet.py (hunk 1 above)
Comment (japokorn) [reply library/blivet.py:417]: This is a workaround. At this moment updating label value to empty string causes blivet to fail. Unfortunately empty string is the default value and it causes failures within storage role tests.

Comment (japokorn) [reply library/blivet.py]: The root cause this error was in another test that tried to modify label of mounted xfs system. Workaround was removed and empty string can be now used to remove existing label.

**Comment by richm** [comment]
please rebase to latest master branch

**Comment by richm** [comment]
[citest pending]

**Comment by richm** [comment]
[citest bad]

**Comment by richm** [comment]
[citest pending]

**Comment by richm** [comment]
[citest bad]

File: library/blivet.py [hunk 2]
```diff
@@ -474,7 +476,14 @@ def _resize(self):
     def _reformat(self):
         """ Schedule actions as needed to ensure the volume is formatted as specified. """
         fmt = self._get_format()
+
         if self._device.format.type == fmt.type:
+            # format is the same, not need to run reformatting
+            dev_label = '' if self._device.format.label is None else self._device.format.label
```
Comment (dwlehman) [inline library/blivet.py:482]: "no need" would be slightly more natural language.

**Review by dwlehman — APPROVED** [review]
An idempotence test might be nice (another invocation w/ the same/current label, then verify `blivet_output.changed` is `False` would work, I think). Otherwise it looks good to me. Thanks.

**Review by richm — APPROVED** [review]
lgtm

**Comment by richm** [comment]
[citest pending]

**Pull request merged**